src/Controller/CompletePayPalOrderAction.php

```diff
@@ -23,25 +26,42 @@ final class CompletePayPalOrderAction
     /** @var OrderProviderInterface */
     private $orderProvider;
 
+    /** @var FactoryInterface */
+    private $stateMachineFactory;
+
+    /** @var ObjectManager */
+    private $orderManager;
+
     public function __construct(
         PaymentStateManagerInterface $paymentStateManager,
         UrlGeneratorInterface $router,
-        OrderProviderInterface $orderProvider
+        OrderProviderInterface $orderProvider,
+        FactoryInterface $stateMachineFactory,
+        ObjectManager $orderManager
     ) {
         $this->paymentStateManager = $paymentStateManager;
         $this->router = $router;
         $this->orderProvider = $orderProvider;
+        $this->stateMachineFactory = $stateMachineFactory;
+        $this->orderManager = $orderManager;
     }
 
     public function __invoke(Request $request): Response
     {
-        $token = (string) $request->attributes->get('token');
-        $order = $this->orderProvider->provideOrderByToken($token);
+        $id = (int) $request->attributes->get('id');
```

Comment:
I think using the `id` could lead to a flow issue here. If I see that in my network, I will try to do the same with `+1` and `+2` and so on.

So, maybe we could stick to the token?

Actually a much better way is to have `token` and `id`, so it's almost impossible to guess.

Reply:
Ok, we cannot use the token, because it's not generated yet. *But* we can use a PayPal order ID which is also quite unique, so would be much safer than the autogenerated ID :)